### PR TITLE
advancedtls: add examples for reloading from file system

### PR DESCRIPTION
**Pull request by ZhenLian**
This PR adds an example in advancedtls on how to reload credentials from file system.

### CODE REVIEW

**Comment by easwars** [comment]
Could you please hold off on this until https://github.com/grpc/grpc-go/pull/3981 gets in. Thanks.

File: security/advancedtls/examples/credential_reloading_from_files/client/main.go
Comment (easwars) [inline security/advancedtls/examples/credential_reloading_from_files/client/main.go:49]: This is the `client` binary. So, we can avoid prefixing all symbol names with `client`.

File: security/advancedtls/examples/credential_reloading_from_files/client/main.go
Comment (easwars) [inline security/advancedtls/examples/credential_reloading_from_files/client/main.go:56]: The `, error` part can be omitted from all logs.

File: security/advancedtls/examples/credential_reloading_from_files/client/main.go
Comment (easwars) [inline security/advancedtls/examples/credential_reloading_from_files/client/main.go:41]: This should use a different name.

File: security/advancedtls/examples/credential_reloading_from_files/server/main.go
Comment (easwars) [inline security/advancedtls/examples/credential_reloading_from_files/server/main.go:60]: Similarly the `server` prefix can be omitted here.

File: security/advancedtls/examples/credential_reloading_from_files/server/main.go
Comment (easwars) [inline security/advancedtls/examples/credential_reloading_from_files/server/main.go:19]: Is this true? I don't see anything about OAuth2 tokens being consumed or validated. Same for the package level comment on the client side.

File: security/advancedtls/examples/credential_reloading_from_files/server/main.go
Comment (easwars) [inline security/advancedtls/examples/credential_reloading_from_files/server/main.go:42]: Single consts or vars don't need to be in blocks.

File: security/advancedtls/examples/credential_reloading_from_files/README.md
Comment (easwars) [inline security/advancedtls/examples/credential_reloading_from_files/README.md:7]: The client is making a single connection to the server. Given that established connections are not affected by change in the files being watched, how does this example demonstrate the credential reloading feature?

File: security/advancedtls/examples/examples_test.sh
Comment (easwars) [inline security/advancedtls/examples/examples_test.sh:1]: I can barely read bash. Please ask someone else to take a look at this.

**Review by easwars — CHANGES_REQUESTED** [review]

File: security/advancedtls/examples/credential_reloading_from_files/client/main.go
Comment (ZhenLian) [reply security/advancedtls/examples/credential_reloading_from_files/client/main.go]: Renamed to `defaultConnTimeout`. Please let me know if it doesn't look right, thanks!

File: security/advancedtls/examples/credential_reloading_from_files/server/main.go
Comment (ZhenLian) [reply security/advancedtls/examples/credential_reloading_from_files/server/main.go]: Ah sorry, this is something I pasted from another example. I've updated the text accordingly.

File: security/advancedtls/examples/credential_reloading_from_files/README.md
Comment (ZhenLian) [reply security/advancedtls/examples/credential_reloading_from_files/README.md]: Gotcha. I understand your concern. I think the purpose of this example is to demonstrate how to set the API. The credential reloading feature is covered in `advancedtls_integration_test`, which reloads credentials in temp files. 
This example is needed by that internal project, which will have a link pointed to this example in their user guide. The original sample code they gave me also didn't include the reloading feature.
I think the current wordings in README is a bit confusing. I've updated some text. Could you please take a look? Thank you so much!

File: security/advancedtls/examples/examples_test.sh
Comment (ZhenLian) [reply security/advancedtls/examples/examples_test.sh:1]: I don't, either :) @dfawley Could you please take a look at this please?
This is just a simple copy from https://github.com/grpc/grpc-go/blob/master/examples/examples_test.sh, with the following minor modifications:
1. put the right test name and output in `EXAMPLES`, `EXPECTED_SERVER_OUTPUT`, and `EXPECTED_CLIENT_OUTPUT`
2. replace `cd ./examples` with `cd ./security/advancedtls/examples`
3. add a `sleep 2s` at the time after we start the server and before we initiate the client request(I added this mainly because otherwise, the test could sometimes fail locally with a server unavailable error. I didn't try on Travis, but when I ran other tests in main grpc it would sometimes fail as well. I am guessing my machine is a bit old that sometimes the server code needs more time to boot up)

File: security/advancedtls/examples/credential_reloading_from_files/client/main.go
Comment (easwars) [inline security/advancedtls/examples/credential_reloading_from_files/client/main.go:42]: Does this have to be this low? Or at least add a comment saying this is set to an impractically low value just for this example, and that in real use cases, this should be in the order of minutes.

**Review by easwars — APPROVED** [review]
The example client and server look good to me. Please also get approval for the bash changes. Thanks.

Also do you think you can add a TODO somewhere (either in the code or in the readme) that this example needs to be updated once https://github.com/grpc/grpc-go/pull/3981 gets in. Again, I'm a little concerned about asking someone to use this example which is going to change very quickly. At least it needs to be clearly communicated to them.

File: security/advancedtls/examples/examples_test.sh
Comment (dfawley) [reply security/advancedtls/examples/examples_test.sh:1]: > add a sleep 2s at the time after we start the server and before we initiate the client request(I added this mainly because otherwise, the test could sometimes fail locally with a server unavailable error. I didn't try on Travis, but when I ran other tests in main grpc it would sometimes fail as well. I am guessing my machine is a bit old that sometimes the server code needs more time to boot up)

That seems strange.  Do you see these errors when running the other `examples_test.sh`?  If the client sends wait-for-ready RPCs then you shouldn't need to worry about this at all.

File: security/advancedtls/examples/credential_reloading_from_files/client/main.go
Comment (ZhenLian) [reply security/advancedtls/examples/credential_reloading_from_files/client/main.go]: Sounds good! It doesn't need to be this low. I've updated it to 1 minute.

File: security/advancedtls/examples/examples_test.sh
Comment (ZhenLian) [reply security/advancedtls/examples/examples_test.sh:1]: Ah, it turned out that the tests failed in other `examples_test.sh` was due to a port is already used on my machine, instead of the unavailable error. So I think other scripts are good.
Just to confirm, the way to send `wait-for-ready` RPCs is through `grpc.WithBlock()`, right? I've updated that, removed the 2s sleep time, and ran the script for 100 times. No errors anymore. 
Could you please check if the script make sense now?

**Comment by ZhenLian** [comment]
@easwars Sounds good. I've added some TODOs here. The other team said they would need the example in a "preview" state, and they've made it clear that things in their user guide could change. So I think it should be fine here.

File: security/advancedtls/examples/examples_test.sh
Comment (dfawley) [reply security/advancedtls/examples/examples_test.sh:1]: > Just to confirm, the way to send wait-for-ready RPCs is through grpc.WithBlock(), right? I've updated that, removed the 2s sleep time, and ran the script for 100 times. No errors anymore.

No, `WithBlock` makes `Dial` wait until there is a ready connection.  `WaitForReady` means the RPC will wait until the channel is ready (however long and after however many failures that takes) before sending the RPC.  It shouldn't be necessary, but if you are seeing weird failures due to connectivity, it can help.  `WithBlock` also shouldn't be necessary; I'd remove it unless you can explain why it fixes things.

Comment (ZhenLian) [reply security/advancedtls/examples/examples_test.sh:1]: OK, sure. I've changed to `WaitForReady` and it also works. Please take a look, thank you so much!

Comment (dfawley) [reply security/advancedtls/examples/examples_test.sh:1]: OH, I thought you wouldn't need WFR either, as both of these options are actually unusual.  But I see there is a race at startup here: if the client `Dial`s before the server is `Listen`ing, then it will get an error.  In this case, either `WithBlock` or WFR RPCs should be fine.

**Review by dfawley — APPROVED** [review]

File: security/advancedtls/examples/examples_test.sh
Comment (ZhenLian) [reply security/advancedtls/examples/examples_test.sh:1]: Great! Thanks for the confirmation!

**Comment by ZhenLian** [comment]
@easwars @dfawley Thanks a lot for the review! Merging now...

**Pull request merged**
